### PR TITLE
Use shared configuration for shortcode- and block-based checkout flows

### LINKED ISSUE #1516: Use shared configuration for Block and Shortcode JS
> Some of these are already available in the JS on the `wc_stripe_params` object, [see here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c182f1dc17614207c5c559abcae220b9735882e/includes/class-wc-gateway-stripe.php#L418-L470).
>
> List of already available things in the format `[this PR] -> [pre-existing]`
>
> - `publicKey -> key`
> - `allowPrepaidCard -> allow_prepaid_card`
> - `inline_cc_form -> inline_cc_form`
>
> I believe all the payment request button options are already available on `wc_stripe_payment_request`, [see here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c182f1dc17614207c5c559abcae220b9735882e/includes/payment-methods/class-wc-stripe-payment-request.php#L507-L548).
>
> I wonder if it'd be better to rely on the pre-existing values — i.e. implicitly trusting that we're not going to change them down the line — or add the new ones here.
>
> I'm personally leaning towards keeping this as is, just to avoid any future problems where we might change the name of a key, remove keys we think have become redundant, etc. I could be convinced otherwise though.
>
> There's one definite bad thing about registering these variables twice; if we change any available filters, remembering to do that in both (eventually more than 2 maybe?) places will be difficult.
>
> Maybe we should have the [main `scripts()` function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c182f1dc17614207c5c559abcae220b9735882e/includes/payment-methods/class-wc-stripe-payment-request.php#L484) detect whether the blocks are loaded, and load in any extras needed for the blocks?
>
> I'd love to hear what you think @nerrad. 
> Same goes for others in Fractal, but I'll ping them in Slack so they don't get pinged for every discussion happening in the PR.
>
> _Originally posted by @reykjalin in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1467#discussion_r601688338_

## TL;DR

- Some configuration we use in the frontend is added twice; once for shortcode and again for blocks.
- We should merge everything that's shared, and add blocks specific configuration in a new variable.

### PR DESCRIPTION
**Pull request by reykjalin**
# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #1516 

1. Refactors the JS configuration preparation into static functions.
    - I'm a little unsure of this approach; is there a better way to make these configurations available to the Blocks Support class?
2. Calls the static JS configuration functions and uses the result for the blocks configuration.
3. Make sure only the Blocks JS is loaded on blocks pages.

# Testing instructions

Rather convoluted; test all the general payment flows, e.g.

- Payment with new card, saved card (Blocks + Shortcode)
- Payment with Payment Request Buttons (PRBs) (Blocks + Shortcode)
- 3DS with both PRBs, new card, saved cards (Blocks + Shortcode)
- Subscription payments (Blocks + Shortcode + Subscriptions >= 3.1.0)
    - Including renewals.
- Pre-Orders (Shortcode only, Blocks are not supported)
- Rendering of default, branded, and custom PRBs.
- Failed payments (Blocks + Shortcode)

Things to watch out for:

- Any sort of JS errors.
- Configurations that don't make it to the front-end (probably associated with JS errors).
- Discrepancies between the Shortcode- and Blocks-based cart/checkout pages.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

File: includes/class-wc-gateway-stripe.php [hunk 1]
```diff
@@ -361,54 +361,12 @@ public function admin_scripts() {
...
-			return;
-		}
-
-		// If no SSL bail.
-		if ( ! $this->testmode && ! is_ssl() ) {
-			WC_Stripe_Logger::log( 'Stripe live mode requires SSL.' );
-			return;
-		}
-
-		$current_theme = wp_get_theme();
```
Comment (dechov) [inline includes/class-wc-gateway-stripe.php:403]: Nice catch – this variable was unused.

File: includes/class-wc-stripe-blocks-support.php [hunk 2]
```diff
@@ -81,26 +93,55 @@ public function get_payment_method_script_handles() {
...
+	 * Returns the Stripe Payment Gateway JavaScript configuration object.
+	 *
+	 * @return array  the JS configuration from the Stripe Payment Gateway.
+	 */
+	private function get_gateway_javascript_configuration() {
+		$js_configuration = [];
+
+		$gateways = WC()->payment_gateways->get_available_payment_gateways();
+		if ( isset( $gateways['stripe'] ) ) {
+			$js_configuration = $gateways['stripe']->javascript_configuration();
```
Comment (dechov) [inline includes/class-wc-stripe-blocks-support.php]: Should these use `$this->name` instead?

File: includes/class-wc-gateway-stripe.php [hunk 3]
```diff
@@ -361,54 +361,12 @@ public function admin_scripts() {
...
-	 *
-	 * Outputs scripts used for stripe payment
+	 * Returns the JavaScript configuration object used on the product, cart, and checkout pages.
 	 *
-	 * @since 3.1.0
-	 * @version 4.0.0
+	 * @return array  The configuration object to be loaded to JS.
 	 */
-	public function payment_scripts() {
+	public function javascript_configuration() {
```
Comment (dechov) [inline includes/class-wc-gateway-stripe.php]: Just a slight preference on my part for something like `js_config` (or `_params` perhaps) – since it's more concise without losing comprehensibility IMO – but no need to change if you'd intentionally named it this way instead.

File: includes/class-wc-gateway-stripe.php [hunk 4]
```diff
@@ -468,7 +426,66 @@ public function payment_scripts() {
...
+			&& ! is_add_payment_method_page()
+			&& ! isset( $_GET['change_payment_method'] ) // wpcs: csrf ok.
+			&& ! ( ! empty( get_query_var( 'view-subscription' ) ) && is_callable( 'WCS_Early_Renewal_Manager::is_early_renewal_via_modal_enabled' ) && WCS_Early_Renewal_Manager::is_early_renewal_via_modal_enabled() )
+			|| ( is_order_received_page() )
+		) {
+			return;
+		}
+
+		// If we're loading the cart or checkout block the scripts here should not be loaded.
+		if ( WC_Stripe_Helper::has_cart_or_checkout_block_on_current_page() ) {
```
Comment (dechov) [inline includes/class-wc-gateway-stripe.php]: As I understand, the idea here is to avoid loading the (shortcode-specific) script when the _shortcode isn't present_, not when a _block is present_. There may be little practical difference (except in extreme edge cases like the cart block sharing a page with checkout shortcode), but looking at the `is_cart` condition [here](https://github.com/woocommerce/woocommerce/blob/0507f06508a448995ae6ebec903115e670cc5ee4/includes/wc-conditional-functions.php#L98), it looks like there's a handy [`wc_post_content_has_shortcode` function](https://github.com/woocommerce/woocommerce/blob/0507f06508a448995ae6ebec903115e670cc5ee4/includes/wc-conditional-functions.php#L412-L422) we can use here.

**Review by dechov — CHANGES_REQUESTED** [review]
Nice improvement! The refactor looks good in general, and I've verified that the "stripe.js" front-end script asset loads only on _shortcode_ cart/checkout, not for the block :+1:

I was originally seeing an error in the Checkout block due to `publicKey` being passed, since we do actually still want the (reverted) change [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1585/commits/6f2f779df93acba93f443d25e73620de6542edd4#diff-dd25ac78a39d9f65aa3b146adc56da769e8cacea23a97d1921b334c17418f563L39-R39).

With that change, payments (incl. 3DS) work with new and saved cards, with both shortcode and blocks. However, Apple Pay in the Cart and Checkout blocks is failing for me, with this error:
> `Unhandled Promise Rejection: ReferenceError: Can't find variable: wc_stripe_payment_request_params`

(I can test some more cases tomorrow.)

**Comment by reykjalin** [comment]
> I was originally seeing an error in the Checkout block due to publicKey being passed, since we do actually still want the (reverted) change here.
>
> With that change, payments (incl. 3DS) work with new and saved cards, with both shortcode and blocks. However, Apple Pay in the Cart and Checkout blocks is failing for me, with this error:
>
> ```
> Unhandled Promise Rejection: ReferenceError: Can't find variable: wc_stripe_payment_request_params
> ```
> (I can test some more cases tomorrow.)

Ugh, sorry about this 🤦 
I must've screwed the revert up while fixing the conflicts that happened during the revert. Sorry about that!

I've pushed a fix in cc9e6e61335ba3bbe3a5924650612abc86f8ec43 and ran some tests to verify that it fixes the issue. Sorry about that!

File: includes/class-wc-stripe-blocks-support.php (hunk 2 above)
Comment (reykjalin) [reply includes/class-wc-stripe-blocks-support.php]: Hmmm, I don't think so because this should be the `id` defined in the gateway, like this one:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/cc9e6e61335ba3bbe3a5924650612abc86f8ec43/includes/class-wc-gateway-stripe.php#L87

And trying to use that becomes sort of circular doesn't it? Instantiating a second gateway to get the ID, only to then use that ID to check for the existence of another gateway?

I agree with your sentiment though; using a string there seems fairly fragile. Maybe we should add a constant with the gateway name somewhere and use that as the ID for the gateway?

Just to be clear; `$this->name` is used to identify the payment method to the Blocks plugin I believe; it's not connected to the Payment Gateway code AFAICT 🤔

File: includes/class-wc-gateway-stripe.php (hunk 3 above)
Comment (reykjalin) [reply includes/class-wc-gateway-stripe.php]: I do prefer to be explicit, which is why I went with the longer name. It's not really that cumbersome to write when you have autocomplete anyway. I'd like to at least keep the `javascript_` part, because `js` is relatively ambiguous in this context IMO.

But on that note, which do you think is the best descriptor for this; `javascript_parameters`/`javascript_params` or `javascript_configuration`/`javascript_config`?

File: includes/class-wc-gateway-stripe.php (hunk 4 above)
Comment (reykjalin) [reply includes/class-wc-gateway-stripe.php]: This is fantastic observation, it really didn't occur to me 🥇 

Made the suggested change in 34935a0f8a8060efa498baa730325167ec2e5411. Awesome catch! 👏 🙌

File: includes/class-wc-gateway-stripe.php (hunk 3 above)
Comment (dechov) [reply includes/class-wc-gateway-stripe.php]: Hm, not a huge deal but I'd vote for `javascript_params` since "configuration" sort of implies to me that it's just user settings when in fact there are also return URLs, nonces, etc., just everything needed for parameterizing the client – what do you think?

File: includes/class-wc-stripe-blocks-support.php (hunk 2 above)
Comment (dechov) [reply includes/class-wc-stripe-blocks-support.php]: Ah, I assumed that the `name` was already bound to be identical to the payment gateway ID for it to work, but I'm not sure if that's true – we can leave this alone 👍

File: includes/class-wc-stripe-helper.php [hunk 5]
```diff
@@ -597,4 +597,25 @@ public static function convert_wc_locale_to_stripe_locale( $wc_locale ) {
...
 		return 'auto';
 	}
+
+	/**
+	 * Checks if this page contains a cart or checkout block.
+	 *
+	 * @since 5.3.0
+	 * @return boolean
+	 */
+	public static function has_cart_or_checkout_block_on_current_page() {
```
Comment (dechov) [inline includes/class-wc-stripe-helper.php]: Should we remove this function since it's no longer being used?

File: includes/class-wc-gateway-stripe.php (hunk 4 above)
Comment (dechov) [reply includes/class-wc-gateway-stripe.php]: Nice – ends up more being concise than I'd expected :+1:

File: includes/class-wc-gateway-stripe.php (hunk 3 above)
Comment (reykjalin) [reply includes/class-wc-gateway-stripe.php]: Your reasoning makes sense, and `javascript_params` sounds good to me 🙂 

Changed the names in c36fa93b5ca73332d2b2400cb2ee4f8abe78db9b 🙌

File: includes/class-wc-stripe-blocks-support.php (hunk 2 above)
Comment (reykjalin) [reply includes/class-wc-stripe-blocks-support.php]: Yeah, we should use the same value for our own sanity, but it's not actually a requirement.

I'll mark this as resolved, but feel free to open the discussion on this again 👍

File: includes/class-wc-stripe-helper.php (hunk 5 above)
Comment (reykjalin) [reply includes/class-wc-stripe-helper.php]: Yes, removed in ed27e97a734d59171aac0c469dfc861533fba865

Nice catch! 👏

File: client/blocks/credit-card/payment-method.js [hunk 6]
```diff
@@ -18,7 +18,7 @@ import { InlineCard, CardElements } from './elements';
  */
 
 export const getStripeCreditCardIcons = () => {
-	return Object.entries( getStripeServerData().icons ).map(
+	return Object.entries( getStripeServerData()?.icons ).map(
```
Comment (dechov) [inline client/blocks/credit-card/payment-method.js]: If there is no data, I think like this particular line would result in an error with or without the optional chaining (and possibly a less helpful one this way)?

File: client/blocks/payment-request/index.js [hunk 7]
```diff
@@ -32,10 +30,9 @@ const paymentRequestPaymentMethod = {
 			return true;
 		}
 
-		// If the `wc_stripe_payment_request_params` object is not available we don't support
-		// payment requests.
-		// eslint-disable-next-line camelcase
-		if ( typeof wc_stripe_payment_request_params === 'undefined' ) {
+		// If the JS Payment Request Button configuration object loaded from PHP is not available
+		// we don't support payment requests.
+		if ( ! getStripeServerData()?.button ) {
```
Comment (dechov) [inline client/blocks/payment-request/index.js]: Is there any known case where this could be false(-y)?

File: client/blocks/stripe-utils/normalize.js [hunk 8]
```diff
@@ -11,6 +9,11 @@
  * @typedef {import('@woocommerce/type-defs/billing').BillingData} CartBillingAddress
  */
 
+/**
+ * Internal dependencies
+ */
+import { getStripeServerData } from '../stripe-utils';
```
Comment (dechov) [inline client/blocks/stripe-utils/normalize.js]: A bit of an odd path, going out of `stripe-utils` and back in – maybe should be `'./utils'` like in the `load-stripe.js` file?

File: client/blocks/stripe-utils/type-defs.js [hunk 9]
```diff
@@ -280,11 +280,9 @@
 /**
  * @typedef {Object} StripeServerData
  *
- * @property {string}                      stripeTotalLabel     The string used for payment
- *                                                              descriptor.
- * @property {string}                      publicKey            The public api key for stripe
+ * @property {string}                      key                  The public api key for stripe
  *                                                              requests.
- * @property {boolean}                     allowPrepaidCard     True means that prepaid cards
+ * @property {boolean}                     allow_prepaid_card   True means that prepaid cards
```
Comment (dechov) [inline client/blocks/stripe-utils/type-defs.js]: I believe this is a `string` as well.

File: includes/class-wc-stripe-blocks-support.php [hunk 10]
```diff
@@ -81,26 +93,55 @@ public function get_payment_method_script_handles() {
...
+	 * Returns the Stripe Payment Gateway JavaScript configuration object.
+	 *
+	 * @return array  the JS configuration from the Stripe Payment Gateway.
+	 */
+	private function get_gateway_javascript_params() {
+		$js_configuration = [];
+
+		$gateways = WC()->payment_gateways->get_available_payment_gateways();
+		if ( isset( $gateways['stripe'] ) ) {
+			$js_configuration = $gateways['stripe']->javascript_params();
```
Comment (dechov) [inline includes/class-wc-stripe-blocks-support.php:126]: Hm, looks like the `get_supported_features` method below just instantiates a `new WC_Gateway_Stripe()`… think we should change one or the other for consistency, or leave them as they are?

**Review by dechov — COMMENTED** [review]
Appears to work well :+1: – I'll do just a bit more testing in a bit.

Minor note: In general, are there known cases in which `getStripeServerData()` will be null, and all the optional chaining would make a difference? Looks like it was added to many of the calls but not all – is there a rule here?

**Comment by reykjalin** [comment]
> In general, are there known cases in which getStripeServerData() will be null, and all the optional chaining would make a difference? Looks like it was added to many of the calls but not all – is there a rule here?

Not any _known_ cases; at least not in terms of a successful flow. That said, Darren mentioned in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1467#discussion_r602778824 on the original Blocks support PR that this function could potentially return `null`, and there's no real guarantee that the shape of the JS object doesn't change down the line.

So I guess you're mostly wondering if it'd be better to have the front-end crash if we change something during development rather than handle the error gracefully without any notification that's what happened?
I'm not sure what's better, but I personally lean towards handling this silently with null coalescing. What do you think?

File: client/blocks/credit-card/payment-method.js (hunk 6 above)
Comment (reykjalin) [reply client/blocks/credit-card/payment-method.js]: Ah yeah, this should definitely be changed to

```js
return Object.entries( getStripeServerData()?.icons ?? {} ).map(
```

> I think like this particular line would result in an error with or without the optional chaining (and possibly a less helpful one this way)?

I think the error isn't helpful in any case to be honest 😅 

image

And if there's an error there that crashes the rest of the Blocks support JS I think. I'd rather use the null coalescing and just add a `?? {}` so that the Stripe blocks support doesn't crash if we happen to forget to supply icons 😬

Comment (reykjalin) [reply client/blocks/credit-card/payment-method.js]: Sorry; just to clarify my point with the error ^ not being useful is that when you don't see where exactly it's coming from — which we don't because webpack minifes the code during the build — that error isn't useful at all.

File: client/blocks/payment-request/index.js (hunk 7 above)
Comment (reykjalin) [reply client/blocks/payment-request/index.js]: There used to be before I added [the `button` prop](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1585/files#diff-3197fd215f7f836c9060fc595ce3eef6b21515846bb890f8d3e0a273b605ce28R109-R111) to the blocks code every time.

Nice catch! We can remove that check 🙂

File: client/blocks/stripe-utils/normalize.js (hunk 8 above)
Comment (reykjalin) [reply client/blocks/stripe-utils/normalize.js]: Yes, I'll change that, nice catch!

File: client/blocks/credit-card/payment-method.js (hunk 6 above)
Comment (reykjalin) [reply client/blocks/credit-card/payment-method.js]: Fixed in eea0676f5d0d0e6bebc7607c271c01ff74bd2d88

File: client/blocks/payment-request/index.js (hunk 7 above)
Comment (reykjalin) [reply client/blocks/payment-request/index.js]: Removed in a62f1744add8409df2f94651ef3297b65614ece6

File: client/blocks/stripe-utils/normalize.js (hunk 8 above)
Comment (reykjalin) [reply client/blocks/stripe-utils/normalize.js]: Fixed in da51b88aa60ddc7a3842e246e32d71e2190e1038

File: client/blocks/stripe-utils/type-defs.js (hunk 9 above)
Comment (reykjalin) [reply client/blocks/stripe-utils/type-defs.js]: Fixed in 677655fb32d043ba61f6cf46bbea39eb845b0699

File: includes/class-wc-stripe-blocks-support.php (hunk 10 above)
Comment (reykjalin) [reply includes/class-wc-stripe-blocks-support.php:126]: Nice catch!

Fixed in 0c9e774c82853a3cfcac3844031f59609ad981a1

File: client/blocks/credit-card/payment-method.js (hunk 6 above)
Comment (dechov) [reply client/blocks/credit-card/payment-method.js]: Sounds good. And I really have no idea how that "like" snuck in to my comment there 😆

**Comment by dechov** [comment]
Well I meant in terms of a _possible_ flow, not necessarily successful. I'm fine with the optional chaining but was mostly wondering why it's in some places and not others, and if there's a pattern there – this example (among others) was left alone: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c9e774c82853a3cfcac3844031f59609ad981a1/client/blocks/credit-card/use-payment-processing.js#L63

**Comment by reykjalin** [comment]
Thanks for calling out the lack of optional chaining Paul!

I've fixed all instances I could find in df49b46d8214f4f0f41ef54f781291529a925c41 🙂

**Review by dechov — APPROVED** [review]
All LGTM 🎉 – tested most of the cases in the testing instructions (though not every single permutation) and some others, and tried various config like PRB settings, inline form, disallowing prepaid cards.
